scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/MetadataJsonValueValidator.scala

```diff
@@ -0,0 +1,392 @@
+        val updatedEntries = entries.map { entry =>
+          val idNel = entry(id)
+
+          idNel match {
+            case Validated.Valid(idValue) =>
+              val idAsString = idValue.str
+              val numOfIdOccurrences = idsThatAppearMoreThanOnce.getOrElse(idAsString, Nil).length
+
+              if numOfIdOccurrences > 1 then entry + (id -> IdIsNotUniqueError(idAsString, s"This id occurs $numOfIdOccurrences times").invalidNel[Value])
+              else entry
+            case invalidNel => entry
+          }
+        }
+        entryType -> updatedEntries
+      }
+    else allEntries
+  }
+
+  def checkIfEntriesHaveCorrectParentIds(
+      allEntries: Map[String, List[ValidatedEntry]],
+      allEntryIds: Map[String, EntryTypeAndParent],
+      entryTypesGrouped: Map[EntryTypeAndParent, List[(String, EntryTypeAndParent)]]
+  ): Map[FieldName, List[ValidatedEntry]] = {
+    def checkIfParentTypeExists(numOfParentTypesToRemove: Int) = {
+      val parentTypes = List("Asset", "ContentFolder", "ArchiveFolder")
+      val allEntryTypes = allEntries.keys
+      val entriesWithAParentType = allEntryTypes.filter(parentTypes.drop(numOfParentTypesToRemove).contains(_))
+      entriesWithAParentType.toSet
+    }
+
+    def getErrorMessageIfSeriesDoesNotExist(entry: ValidatedEntry, entryType: String) = {
+      if entry.contains(series) && entryType == "File" then Map(series -> SeriesExistsError("A file can not have a Series").invalidNel[Value])
+      else if entry.contains(series) || entryType == "File" then Map()
+      else
+        Map(
+          series -> SeriesDoesNotExistError(
+            "The parentId is null and since only top-level entries can have a null parentId " +
+              "and a series, this entry should have a 'series' (if it is indeed top-level)"
+          ).invalidNel[Value]
+        )
+    }
+
+    def getErrorMessageIfASeriesExists(entry: ValidatedEntry) = {
+      if entry.contains(series) then
+        Map(
+          series -> SeriesExistsError(
+            "This entry has a series but has a parentId that's not null; only a top-level entry can have this"
+          ).invalidNel[Value]
+        )
+      else Map()
+    }
+
+    allEntries.map { case (entryType, entriesOfSpecificType) =>
+      val updatedEntries =
+        entriesOfSpecificType.map { entry =>
+          val idNel = entry(id)
+          idNel match {
+            case Validated.Valid(idValue) =>
+              val idAsString = idValue.str
+              val entryTypeAndParent = allEntryIds(idAsString)
+              val parentIdNel = entry(parentId)
+
+              val (validatedParentId, potentialSeriesError) =
+                entryTypeAndParent.potentialParentId match {
+                  case None =>
+                    val parentIdValidatedNel = entryTypeAndParent match {
+                      // It's hard to determine whether an ArchiveFolderEntry's parent is null/None because it's a top-level entry or it's a mistake
```

Comment:
Also, regarding using a tuple, I did it because, I'm returning both `parentId` errors and  `series` errors, so this line `val entryWithUpdatedParentId = entry + (parentId -> validatedParentId)` is not correct, but what I could do is make `validatedParentId` a Map that contains `parentId` and  `series` and change the name of the variable.

Reply:
In addition to the other two comments, can a file have a `null` parentId? I thought that only `Asset`, `ContentFolder` and `ArchiveFolder` could.

Reply:
No, that's right, a file can't have a null parent id.